packages/binding-coap/src/mdns-introducer.ts

```diff
@@ -0,0 +1,142 @@
+
+    private readonly ipAddressFamily: IpAddressFamily;
+
+    constructor(address?: string, ipAddressFamily?: IpAddressFamily) {
+        this.ipAddressFamily = ipAddressFamily ?? "IPv4";
+
+        const type = ipAddressFamily === "IPv6" ? "udp6" : "udp4";
+
+        this.mdns = makeMdns({
+            ip: address,
+            type,
+        });
+        this.mdnsEntries = new Map();
+        this.mdns.on("query", (query) => {
+            this.sendMdnsResponses(query);
+        });
+    }
+
+    private sendMdnsResponses(query: makeMdns.QueryPacket): void {
+        this.mdnsEntries.forEach((value) => {
+            const entryName = value[0].name;
+            const matchingQuestions = query.questions.filter((question) => question.name === entryName);
+
+            if (matchingQuestions.length <= 0) {
+                return;
+            }
+
+            this.mdns.respond(value);
+        });
+    }
+
+    private determineTarget(): string {
+        const interfaces = networkInterfaces();
+
+        for (const iface in interfaces) {
+            for (const entry of interfaces[iface] ?? []) {
+                if (entry.internal === false) {
+                    if (entry.family === this.ipAddressFamily) {
+                        return entry.address;
+                    }
+                }
+            }
+        }
+
+        // TODO: Is it correct to throw an error here?
+        throw Error("Found no suitable IP address for performing MDNS introduction.");
+    }
+
+    private createTxtData(parameters: MdnsDiscoveryParameters): Array<string> {
+        const txtData = [`td=${parameters.urlPath}`];
+
+        const type = parameters.type;
+        if (type != null) {
+            txtData.push(`type=${type}`);
+        }
+
+        const scheme = parameters.scheme;
+        if (scheme != null) {
+            txtData.push(`scheme=${scheme}`);
+        }
+
+        return txtData;
+    }
+
+    public registerExposedThing(thing: ExposedThing, parameters: MdnsDiscoveryParameters): void {
+        const serviceName = parameters.serviceName;
+        const instanceName = `${thing.title}.${serviceName}`;
+
+        const target = this.determineTarget();
+        const txtData = this.createTxtData(parameters);
+
+        this.mdnsEntries.set(parameters.urlPath, [
+            {
+                name: serviceName,
+                type: "PTR",
+                data: instanceName,
+            },
+            {
+                name: instanceName,
+                type: "SRV",
+                data: {
+                    port: parameters.port,
+                    target,
+                },
+            },
+            {
+                name: instanceName,
+                type: "TXT",
+                data: txtData,
+            },
+        ]);
+    }
+
+    public delete(urlPath: string): void {
+        this.mdnsEntries.delete(urlPath);
+    }
+
+    public close(): void {
+        this.mdns.destroy();
```

Comment:
Watch out, the destroy method has an undocumented signature: https://github.com/mafintosh/multicast-dns/blob/master/index.js#L114 (note also that the callback is passed down to `socket.close`). I think we should refactor this to a Promise to avoid dangling resources.

Reply:
Thank you for the hint! I hope the implementation is now improved :)